File: codebase/core/main/src/AxleModule/ForestModule.fs
namespace Forest
open System
open System.Reflection
open System.Runtime.CompilerServices
open Axle
open Axle.DependencyInjection
open Axle.Logging
open Axle.Modularity
open Axle.Verification
open Forest
open Forest.ComponentModel
open Forest.Engine
open Forest.Engine.Aspects
open Forest.Security
open Forest.StateManagement
open Forest.Templates
open Forest.UI
open System.Diagnostics

type [<Interface>] IForestIntegrationProvider =
    abstract member Renderer : IPhysicalViewRenderer with get
    abstract member StateProvider : IForestStateProvider with get

[<AttributeUsage(AttributeTargets.Class|||AttributeTargets.Interface, Inherited = true, AllowMultiple = false)>]
type [<Sealed>] RequiresForestAttribute() = inherit RequiresAttribute(typeof<ForestModule>)

and [<Interface;Module;RequiresForest>] IForestViewProvider =
    abstract member RegisterViews: registry : IViewRegistry -> unit

and [<Sealed;NoEquality;NoComparison;Module;Requires(typeof<ForestTemplatesModule>)>] 
    internal ForestModule(
        container : IContainer, 
        templateProvider : ITemplateProvider, 
        app : Application, 
        rtp : ResourceTemplateProvider, 
        logger : ILogger) =
    [<DefaultValue>]
    val mutable private _context : IForestContext
    [<DefaultValue>]
    val mutable private _integrationProvider : IForestIntegrationProvider voption
    //[<DefaultValue>]
    //val mutable private _forestEngine : IForestEngine

    //member this.InitForest() =
        //let pvr, sp =
        //    match this._integrationProvider with
        //    | ValueSome ip ->
        //        ip.Renderer, ip.StateProvider
        //    | ValueNone ->
        //        (NoOp.PhysicalViewRenderer() :> IPhysicalViewRenderer, DefaultForestStateProvider() :> IForestStateProvider)
        //this._forestEngine <- ForestEngine.Create this._context sp pvr logger
    
    [<ModuleInit>]
    member this.Init(e : ModuleExporter) =
        let securityManager =
            match container.TryResolve<ISecurityManager>() with
            | (true, sm) -> sm
            | (false, _) -> upcast NoopSecurityManager()
        let viewFactory =
            match null2vopt container.Parent with
            | ValueSome c -> (c, app)
            | ValueNone -> (container, app)
            |> ContainerViewFactory
        let context : IForestContext = upcast DefaultForestContext(viewFactory, securityManager, templateProvider, seq { yield this :> IForestExecutionAspect })
        this._context <- context
        //this.InitForest()
        context |> e.Export<IForestContext> |> ignore
        this |> e.Export<IForestEngine> |> ignore


    [<ModuleDependencyInitialized>]
    member this.DependencyInitialized (viewProvider : IForestViewProvider) =
        (this :> IViewRegistry) |> viewProvider.RegisterViews

    [<ModuleDependencyInitialized>]
    member this.DependencyInitialized (integration : IForestIntegrationProvider) =
        match this._integrationProvider with
        | ValueNone ->
            this._integrationProvider <- ValueSome integration
        | ValueSome _ ->
            invalidOp "Forest integration is already configured"
        //this.InitForest()

    //member private this.ForestEngine with get() = this._forestEngine   
    
    member private this.CreateEngine() =
        let pvr, sp =
            match this._integrationProvider with
            | ValueSome ip ->
                ip.Renderer, ip.StateProvider
            | ValueNone ->
                (NoOp.PhysicalViewRenderer() :> IPhysicalViewRenderer, DefaultForestStateProvider() :> IForestStateProvider)
        upcast new ForestEngine(this._context, sp, pvr) : IForestEngine

    interface IForestEngine with
        [<Obsolete>]
        member this.RegisterSystemView<'sv when 'sv :> ISystemView>() = 
            //this.ForestEngine.RegisterSystemView<'sv>()
            this.CreateEngine().RegisterSystemView<'sv>()

    interface ITreeNavigator with
        member this.Navigate t = 
            //this.ForestEngine.Navigate t
            this.CreateEngine().Navigate t

        member this.Navigate (t, m) = 
            //this.ForestEngine.Navigate (t, m)
            this.CreateEngine().Navigate (t, m)

    interface IMessageDispatcher with
        member this.SendMessage msg = 
            //this.ForestEngine.SendMessage msg
            this.CreateEngine().SendMessage (msg)

    interface ICommandDispatcher with
        member this.ExecuteCommand (cmd, target, arg) = 
            //this.ForestEngine.ExecuteCommand (cmd, target, arg)
            this.CreateEngine().ExecuteCommand (cmd, target, arg)

    interface IViewRegistry with
        member this.GetDescriptor(viewType : Type) : IViewDescriptor = this._context.ViewRegistry.GetDescriptor viewType
        member this.GetDescriptor(name : vname) : IViewDescriptor = this._context.ViewRegistry.GetDescriptor name
        member this.Register<'t when 't:>IView>() :IViewRegistry = 
            typeof<'t>.GetTypeInfo().Assembly |> rtp.RegisterAssemblySource 
            this._context.ViewRegistry.Register<'t>()
        member this.Register(t : Type) : IViewRegistry = 
            t.GetTypeInfo().Assembly |> rtp.RegisterAssemblySource 
            this._context.ViewRegistry.Register t
    interface IViewFactory with
        member this.Resolve(descriptor : IViewDescriptor): IView = this._context.ViewFactory.Resolve descriptor
        member this.Resolve(descriptor : IViewDescriptor, model : obj): IView = this._context.ViewFactory.Resolve(descriptor, model)

    interface IForestExecutionAspect with
        member __.Navigate(cutPoint: IForestExecutionCutPoint) =
            let sw = Stopwatch.StartNew()
            cutPoint.Proceed()
            sw.Stop()
            logger.Trace("Forest Navigate operation took {0}ms", sw.ElapsedMilliseconds)
        member __.ExecuteCommand(cutPoint: IForestExecutionCutPoint) =
            let sw = Stopwatch.StartNew()
            cutPoint.Proceed()
            sw.Stop()
            logger.Trace("Forest ExecuteCommand operation took {0}ms", sw.ElapsedMilliseconds)
        member __.SendMessage(cutPoint: IForestExecutionCutPoint) =
            let sw = Stopwatch.StartNew()
            cutPoint.Proceed()
            sw.Stop()
            logger.Trace("Forest SendMessage operation took {0}ms", sw.ElapsedMilliseconds)

[<Extension>]
type Extensions =
    [<Extension>]
    static member UseForest(NotNull "builder" builder : Axle.IApplicationBuilder) =
        builder.ConfigureModules(fun m -> m.Load(typeof<ForestModule>) |> ignore)

File: codebase/core/main/src/ForestModule.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using Axle.DependencyInjection;
using Axle.Logging;
using Axle.Modularity;
using Forest.ComponentModel;
using Forest.Dom;
using Forest.Engine;
using Forest.Engine.Aspects;
using Forest.Engine.Instructions;
using Forest.Globalization;
using Forest.Navigation;
using Forest.Security;
using Forest.StateManagement;
using Forest.Templates;

namespace Forest
{
    [Module]
    [Requires(typeof(ForestViewRegistry))]
    [Requires(typeof(ForestTemplatesModule))]
    [Requires(typeof(ForestNavigationModule))]
    [Requires(typeof(ForestGlobalizationModule))]
    [Requires(typeof(ForestSecurityModule))]
    internal sealed class ForestModule :
        IForestEngine, 
        IForestViewFactory, 
        IForestContext, 
        IForestCommandAdvice, 
        IForestMessageAdvice, 
        IForestNavigationAdvice
    {
        private readonly IForestViewRegistry _viewRegistry;
        private readonly IForestViewFactory _viewFactory;
        private readonly IForestSecurityManager _securityManager;
        private readonly IForestSecurityExceptionHandler _securityExceptionHandler;
        private readonly ITemplateProvider _templateProvider;
        private readonly IForestDomManager _domManager;
        private readonly ICollection<IForestCommandAdvice> _commandAdvices;
        private readonly ICollection<IForestMessageAdvice> _messageAdvices;
        private readonly ICollection<IForestNavigationAdvice> _navigationAdvices;
        private readonly ILogger _logger;
        private readonly ForestGlobalizationModule _forestGlobalizationModule;
        private ForestEngineContextProvider _engineContextProvider;

        public ForestModule(
                IForestViewRegistry viewRegistry, 
                IDependencyContext dependencyContainer, 
                ITemplateProvider templateProvider,
                IForestSecurityManager securityManager,
                IForestSecurityExceptionHandler securityExceptionHandler,
                IDependencyContainerFactory dependencyContainerFactory, 
                ForestGlobalizationModule globalizationModule,
                ILogger logger)
        {
            _viewRegistry = viewRegistry;
            _viewFactory = new ContainerViewFactory(dependencyContainer.Parent ?? dependencyContainer, dependencyContainerFactory);
            _securityManager = securityManager;
            _securityExceptionHandler = securityExceptionHandler;
            _templateProvider = templateProvider;
            _domManager = new ForestDomManager(this, logger);
            _messageAdvices = new List<IForestMessageAdvice>();
            _commandAdvices = new List<IForestCommandAdvice>();
            _navigationAdvices = new List<IForestNavigationAdvice>();
            _forestGlobalizationModule = globalizationModule;
            _logger = logger;
        }

        [ModuleInit]
        internal void Initialize(IDependencyExporter exporter)
        {
            _messageAdvices.Add(this);
            _commandAdvices.Add(this);
            _navigationAdvices.Add(this);
            exporter
                .Export(this)
                .Export(_viewRegistry)
                .Export(_securityExceptionHandler)
                .Export<IForestStateInspector>(new DefaultForestStateInspector())
                ;
        }

        [ModuleDependencyInitialized]
        internal void DependencyInitialized(ForestEngineContextProvider engineContextProvider)
        {
            if (_engineContextProvider != null)
            {
                throw new InvalidOperationException("Forest engine provider is already configured");
            }
            _engineContextProvider = engineContextProvider;
        }

        [ModuleDependencyInitialized]
        internal void DependencyInitialized(IForestCommandAdvice forestCommandAdvice) => _commandAdvices.Add(forestCommandAdvice);
        [ModuleDependencyInitialized]
        internal void DependencyInitialized(IForestMessageAdvice forestMessageAdvice) => _messageAdvices.Add(forestMessageAdvice);
        [ModuleDependencyInitialized]
        internal void DependencyInitialized(IForestNavigationAdvice forestNavigationAdvice) => _navigationAdvices.Add(forestNavigationAdvice);

        [ModuleDependencyTerminated]
        internal void DependencyTerminated(IForestCommandAdvice forestCommandAdvice) => _commandAdvices.Remove(forestCommandAdvice);
        [ModuleDependencyTerminated]
        internal void DependencyTerminated(IForestMessageAdvice forestMessageAdvice) => _messageAdvices.Remove(forestMessageAdvice);
        [ModuleDependencyTerminated]
        internal void DependencyTerminated(IForestNavigationAdvice forestNavigationAdvice) => _navigationAdvices.Remove(forestNavigationAdvice);

        internal ForestEngineContextProvider EngineContextProvider => _engineContextProvider;

        T IForestEngine.RegisterSystemView<T>()
        {
            using (var ctx = EngineContextProvider.GetContext(this, SystemViewDescriptors, _logger))
            {
                return ctx.Engine.RegisterSystemView<T>();
            }
        }
        IView IForestEngine.RegisterSystemView(Type viewType)
        {
            using (var ctx = EngineContextProvider.GetContext(this, SystemViewDescriptors, _logger))
            {
                return ctx.Engine.RegisterSystemView(viewType);
            }
        }
        void ITreeNavigator.Navigate(Location location)
        {
            using (var ctx = EngineContextProvider.GetContext(this, SystemViewDescriptors, _logger))
            {
                ctx.Engine.Navigate(location);
            }
        }
        void ITreeNavigator.NavigateBack()
        {
            using (var ctx = EngineContextProvider.GetContext(this, SystemViewDescriptors, _logger))
            {
                ctx.Engine.NavigateBack();
            }
        }
        void ITreeNavigator.NavigateBack(int offset)
        {
            using (var ctx = EngineContextProvider.GetContext(this, SystemViewDescriptors, _logger))
            {
                ctx.Engine.NavigateBack(offset);
            }
        }
        void ITreeNavigator.NavigateUp()
        {
            using (var ctx = EngineContextProvider.GetContext(this, SystemViewDescriptors, _logger))
            {
                ctx.Engine.NavigateUp();
            }
        }
        void ITreeNavigator.NavigateUp(int offset)
        {
            using (var ctx = EngineContextProvider.GetContext(this, SystemViewDescriptors, _logger))
            {
                ctx.Engine.NavigateUp(offset);
            }
        }
        
        void ICommandDispatcher.ExecuteCommand(string command, string target, object arg)
        {
            using (var ctx = EngineContextProvider.GetContext(this, SystemViewDescriptors, _logger))
            {
                ctx.Engine.ExecuteCommand(command, target, arg);
            }
        }

        void IMessageDispatcher.SendMessage<T>(T msg)
        {
            using (var ctx = EngineContextProvider.GetContext(this, SystemViewDescriptors, _logger))
            {
                ctx.Engine.SendMessage(msg);
            }
        }
        IView IForestViewFactory.Resolve(IForestViewDescriptor descriptor) => _viewFactory.Resolve(descriptor);
        IView IForestViewFactory.Resolve(IForestViewDescriptor descriptor, object model) => _viewFactory.Resolve(descriptor, model);
        
        IForestViewFactory IForestContext.ViewFactory => this;
        IForestViewRegistry IForestContext.ViewRegistry => _viewRegistry;
        IForestSecurityManager IForestContext.SecurityManager => _securityManager;
        ITemplateProvider IForestContext.TemplateProvider => _templateProvider;
        IForestDomManager IForestContext.DomManager => _domManager;
        IDomProcessor IForestContext.GlobalizationDomProcessor => _forestGlobalizationModule;
        IEnumerable<IForestCommandAdvice> IForestContext.CommandAdvices => _commandAdvices;
        IEnumerable<IForestMessageAdvice> IForestContext.MessageAdvices => _messageAdvices;
        IEnumerable<IForestNavigationAdvice> IForestContext.NavigationAdvices => _navigationAdvices;

        bool IForestCommandAdvice.ExecuteCommand(IExecuteCommandPointcut pointcut)
        {
            try
            {
                var sw = Stopwatch.StartNew();
                var result = pointcut.Proceed();
                sw.Stop();
                _logger.Info("Forest ExecuteCommand operation took {0}ms", sw.ElapsedMilliseconds.ToString());
                return result;
            }
            catch (ForestSecurityException securityException)
            {
                _securityExceptionHandler.HandleSecurityException(securityException, pointcut.Command, this);
            }
            return false;
        }

        bool IForestMessageAdvice.SendMessage(ISendMessagePointcut pointcut)
        {
            try
            {
                var sw = Stopwatch.StartNew();
                var result = pointcut.Proceed();
                sw.Stop();
                _logger.Info("Forest SendMessage operation took {0}ms", sw.ElapsedMilliseconds.ToString());
                return result;
            }
            catch (ForestSecurityException securityException)
            {
                _securityExceptionHandler.HandleSecurityException(securityException, this);
            }
            return false;
        }

        bool IForestNavigationAdvice.Navigate(INavigatePointcut pointcut)
        {
            try
            {
                var sw = Stopwatch.StartNew();
                var result = pointcut.Proceed();
                sw.Stop();
                _logger.Info("Forest Navigate operation took {0}ms", sw.ElapsedMilliseconds.ToString());
                return result;
            }
            catch (ForestSecurityException securityException)
            {
                _securityExceptionHandler.HandleSecurityException(securityException, pointcut.Location, this);
            }
            return false;
        }

        internal IEnumerable<IForestViewDescriptor> SystemViewDescriptors => _viewRegistry.ViewDescriptors.Where(x => x.IsSystemView);
    }
}
